Integrate Rules Engine

hobinjk opened the pull request:
This is an alternate take on #252 which additionally incorporates all of the Rules Engine's code.

Todo:
- [x] Incorporate rules engine into experiments list
- [x] Redirect unauthenticated users to '/login'

benfrancis commented:
Nice, this is much easier to get running :)

![thunderdome](https://user-images.githubusercontent.com/552417/29966048-09cedbee-8f08-11e7-947b-587599ed2cd0.gif)

I think integrating it like this could simplify a lot of things until we come up with an add-ons system to make the codebase more modular. Are you OK with that?

FWIW when I run this I'm still getting errors in the console:
> SyntaxError: expected expression, got '<'  Rule.js:1
> ReferenceError: Rule is not defined  edit.js:107:3

and I can't get drag and drop to work.

hobinjk commented:
Two repos enter, one repo leaves!

I'm definitely okay with using this integrated approach.

From that error message it looks like `Rule.js` is missing. Typing that out made me realize that this is a bug caused by filesystem case-sensitivity which I can fix right now.

benfrancis commented:
@hobinjk Am I right in thinking this is still WIP? I'm confused by the review flags. Just want to make sure I'm not blocking you.

hobinjk commented:
@benfrancis Sorry for the confusion! This is ready for review as of two seconds ago

hobinjk commented on src/app.js:
```diff
@@ -118,6 +124,23 @@ function getOptions() {
   return options;
 }
 
+/**
```
@benfrancis picking up from #268. `rulesEngineConfigure` exists because `Router.configure` runs before we know whether the server will be `http` or `https`. If we're running the tests then we also don't know the server's port. This is a problem because `node-fetch` only works with a complete URL.

It's unnecessary for this function to also issue the JWT. The `rules-engine/index.js` file can use JSONWebToken directly pretty easily. I'm going to switch it over to that approach so it's more clear why this code exists.

benfrancis commented on static/rules-engine/index.css:
```diff
@@ -0,0 +1,182 @@
+html, body {
+  margin: 0;
+  padding: 0;
```
For the rest of the gateway we set a base 10px font size and then use rem rather than px for sizing everything. The reason for this is to make it easier to adjust to screens of different pixel densities by just changing the base font size. Currently all the rules engine dimensions are set in px and it doesn't scale well on mobile.

hobinjk commented:
Thanks for the review! This PR is certainly a bit on the large size.

I think that full integration should be left to a future PR if we decide to go for it.

I will put the front end improvements in this PR

hobinjk commented on src/app.js:
```diff
@@ -118,6 +124,23 @@ function getOptions() {
+ * the gateway needs to configure it with a JWT and a server address
+ * @param {http.Server|https.Server} server
+ */
+function rulesEngineConfigure(server) {
```
I left a comment above with a bit of the context of this. Where would you put the localBaseUrl? I think having a server to baseURL function would eliminate most of the complexity of this. `JSONWebToken.requestToken(-1)` is the `requestJWT` method.

There's definitely the potential for a bit of duplication. I think that having a wrapper of `node-fetch` that automatically acquires a JWT and is configured with the base URL would help considerably. It would eliminate the need for the intent parser to do anything fancy and would eliminate some amount of the Rules Engine's fetch code.

hobinjk commented on src/constants.js:
```diff
@@ -21,6 +21,7 @@ exports.LOG_OUT_PATH = '/log-out';
 exports.SETTINGS_PATH = '/settings';
 exports.UPLOADS_PATH = '/uploads';
 exports.DEBUG_PATH = '/debug';
+exports.RULES_ENGINE_PATH = '/rules-engine';
```
Now that it's integrated, the path can certainly be made more consistent

hobinjk commented on src/router.js:
```diff
@@ -48,6 +48,10 @@ var Router = {
 
+    let rulesEngine = require('./rules-engine/index.js');
+    app.use(APP_PREFIX + Constants.RULES_ENGINE_PATH, rulesEngine);
+    app.use(API_PREFIX + Constants.RULES_ENGINE_PATH, auth, rulesEngine);
```
Yes, the problem is that the rules engine HTML is not part of the single page app.

hobinjk commented on src/rules-engine/.eslintrc.js:
```diff
@@ -0,0 +1,29 @@
+module.exports = {
```
Nope, I'll delete both

hobinjk commented on src/rules-engine/Database.js:
```diff
@@ -0,0 +1,115 @@
+/**
+ * Open the database
+ */
+Database.prototype.open = function() {
```
I'll build this on top of `db.js` to share as much functionality as possible

hobinjk commented on src/rules-engine/actions/Action.js:
```diff
@@ -0,0 +1,41 @@
+/**
```
I can call them Effects, names have power :)

hobinjk commented:
Putting the rules and rule views into the SPA architecture was simpler than expected, so I've updated it accordingly. I believe that I've addressed most of the review comments except for scaling the UI down on mobile. It's usable on a simulated larger phone/tablet but it just runs out of available space on smaller devices.

benfrancis commented on src/constants.js:
```diff
@@ -21,6 +21,7 @@ exports.LOG_OUT_PATH = '/log-out';
 exports.SETTINGS_PATH = '/settings';
 exports.UPLOADS_PATH = '/uploads';
 exports.DEBUG_PATH = '/debug';
+exports.RULES_ENGINE_PATH = '/rules';
```
\o/

nit: This could just be called RULES_PATH for consistency.

hobinjk commented:
Thanks! I've resolved the first two bugs and renamed RULES_ENGINE_PATH. I'm going to leave the mobile UX improvements to a future PR.
